gluten-core/src/main/scala/org/apache/gluten/extension/columnar/OffloadSingleNode.scala

```diff
@@ -265,13 +265,35 @@ case class OffloadProject() extends OffloadSingleNode with LogLevelUtil {
       newProjectList
     }
 
+    def replaceInputFileAttr(namedExpr: NamedExpression): NamedExpression = namedExpr.name match {
+      case "input_file_name" =>
+        Alias(InputFileName(), namedExpr.name)(namedExpr.exprId)
+      case "input_file_block_start" =>
+        Alias(InputFileBlockStart(), namedExpr.name)(namedExpr.exprId)
+      case "input_file_block_length" =>
+        Alias(InputFileBlockLength(), namedExpr.name)(namedExpr.exprId)
+      case _ => namedExpr
+    }
+
     plan match {
       case f: FileSourceScanExec =>
-        f.copy(output = genNewOutput(f.output))
+        if (FallbackTags.nonEmpty(f)) {
+          val p = ProjectExec(genNewOutput(f.output).map(replaceInputFileAttr), f)
+          FallbackTags.add(p, "fallback input file expression")
+          p
+        } else {
+          f.copy(output = genNewOutput(f.output))
+        }
       case f: FileSourceScanExecTransformer =>
         f.copy(output = genNewOutput(f.output))
       case b: BatchScanExec =>
-        b.copy(output = genNewOutput(b.output).asInstanceOf[Seq[AttributeReference]])
+        if (FallbackTags.nonEmpty(b)) {
+          val p = ProjectExec(genNewOutput(b.output).map(replaceInputFileAttr), b)
+          FallbackTags.add(p, "fallback input file expression")
+          p
+        } else {
+          b.copy(output = genNewOutput(b.output).asInstanceOf[Seq[AttributeReference]])
+        }
       case b: BatchScanExecTransformer =>
```

Comment:
Is there real case that we see a `BatchScanExecTransformer` in `OffloadProject`? Since `OffloadOthers` is executed after `OffloadProject`.